### PR TITLE
#18 split join

### PR DESCRIPTION
**Pull request by TONYSKYZENG**

### CODE REVIEW

File: include/WindowSlider/SplitJoinWS.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  \li Configure window length: @ref setWindowLen
+  \li Configure slide length: @ref setSlideLen (default is 1 if not called)
+  \li Set parallel executing behavior on SMP,@ref setParallelSMP
+  \li Optional, (@ref setRunTimeScheduling)
+  \li To make the parallel join processors started, @ref initJoinProcessors
+  \li Feed tuples @ref feedTupleS or @ref feedTupleR
+  \li Terminate, by @ref terminateJoinProcessors
+*
+*/
+class SplitJoinWS : public AbstractWS, public AbstractC20Thread {
```
Comment (ShuhaoZhangTony) [inline include/WindowSlider/SplitJoinWS.h:32]: The name is werid. AbstractC20Thread?

File: src/JoinAlgo/CMakeLists.txt [hunk 2]
```diff
@@ -1,6 +1,7 @@
 add_sources(
         JoinAlgoTable.cpp
         AbstractJoinAlgo.cpp
+        NestedLoopJoin.cpp
```
Comment (ShuhaoZhangTony) [inline src/JoinAlgo/CMakeLists.txt:4]: ??NLJ? Oh, the base algorithm from SplitJoin Paper.

Comment (ShuhaoZhangTony) [reply src/JoinAlgo/CMakeLists.txt:4]: Actually, you can use this as the single-thread verification algorithm.

File: src/JoinAlgo/NestedLoopJoin.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  size_t result = 0;
+  for (size_t i = 0; i < tsLen; i++) {
+    if (ts[i]->key == tr->key) {
+      result++;
+    }
+  }
+  // cout<<result<<endl;
+  return result;
+}
+size_t NestedLoopJoin::join(TuplePtr *ts, TuplePtr *tr, size_t tsLen, size_t trLen, int threads) {
```
Comment (ShuhaoZhangTony) [inline src/JoinAlgo/NestedLoopJoin.cpp:18]: Shared Pointer with Pointer?

File: src/JoinProcessor/SplitJoinJP.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+// Created by tony on 18/03/22.
+//
+
+#include <JoinProcessor/SplitJoinJP.h>
+using namespace INTELLI;
+void SplitJoinJP::joinS(TuplePtr ts) {
+  size_t timeNow = ts->subKey;
+  expireR(timeNow);
+  //single thread join window r and tuple s
+  TuplePtrQueueLocalS->push(ts);
```
Comment (ShuhaoZhangTony) [inline src/JoinProcessor/SplitJoinJP.cpp:11]: Is this the local storage of each joiner?

Comment (ShuhaoZhangTony) [reply src/JoinProcessor/SplitJoinJP.cpp:11]: If so, add some more comments.

**Review by ShuhaoZhangTony — CHANGES_REQUESTED** [review]

File: include/WindowSlider/SplitJoinWS.h (hunk 1 above)
Comment (TONYSKYZENG) [reply include/WindowSlider/SplitJoinWS.h:32]: A basic thread class under C20 style, see Utils

File: src/JoinAlgo/NestedLoopJoin.cpp (hunk 3 above)
Comment (TONYSKYZENG) [reply src/JoinAlgo/NestedLoopJoin.cpp:18]: Use it as an array, legacy way for the ease of partition,etc.

File: src/JoinProcessor/SplitJoinJP.cpp (hunk 4 above)
Comment (TONYSKYZENG) [reply src/JoinProcessor/SplitJoinJP.cpp:11]: Do this tmr

File: src/JoinAlgo/NestedLoopJoin.cpp (hunk 3 above)
Comment (ShuhaoZhangTony) [reply src/JoinAlgo/NestedLoopJoin.cpp:18]: Don't do this. Use the vector class to represent array to avoid explicit pointer.

Comment (TONYSKYZENG) [reply src/JoinAlgo/NestedLoopJoin.cpp:18]: We can not simply replace this with vector. Talk about it this afternoon..

File: benchmark/src/Benchmark.cpp [hunk 5]
```diff
@@ -41,29 +42,38 @@ int main() {
...
+  ewj.test(joinResult, relationCouple);
   //Print result number
   // INTELLI::UtilityFunctions::timerEnd(joinResult);
   joinResult.statPrinter();
+
   dataSet.load3VText(relationCouple.relationR, fileRName);
   dataSet.load3VText(relationCouple.relationS, fileSName);
   joinResult.streamSize = relationCouple.relationR.size();
-  AbstractJoinMethod<AbstractEagerWS> ewj;
+  AbstractJoinMethod<SplitJoinWS> ewj2;
```
[the remark this answers is not in the document]
Comment (TONYSKYZENG) [reply benchmark/src/Benchmark.cpp:60]: commented with/**/

File: src/JoinProcessor/SplitJoinJP.cpp (hunk 4 above)
Comment (TONYSKYZENG) [reply src/JoinProcessor/SplitJoinJP.cpp:11]: See the accordingly *.h

**Pull request merged**